### PR TITLE
Ray cluster CRD and example CR + multi-ray-cluster operator

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
## Why are these changes needed?

Following up on #11929, this PR adds a k8s CRD describing a ray cluster configuration and an example ray cluster CR. A ray cluster CR is pretty much just a reformatted version of one of the current ray cluster configs. 

This PR also extends the operator such that it can manage multiple Ray clusters.
Using kubectl, users can create/update/delete clusters and check monitoring logs.

---

## Related issue number
#11929 #11545

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

Create, update, delete, logging working as expected for me locally.

### CODE REVIEW

File: python/ray/autoscaler/kubernetes/operator_configs/example_cluster_resource.yaml [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+          imagePullPolicy: Always
+          image: rayproject/ray:nightly
+          # Do not change this command - it keeps the pod alive until it is
+          # explicitly killed.
+          command: ["/bin/bash", "-c", "--"]
+          args: ['trap : TERM INT; sleep infinity & wait;']
+          ports:
+          - containerPort: 6379 # Redis port.
+          - containerPort: 6380 # Redis port.
+          - containerPort: 6381 # Redis port.
```
Comment (yiranwang52) [inline python/ray/autoscaler/kubernetes/operator_configs/example_cluster_resource.yaml:40]: do we still need 3 redis ports?

File: python/ray/autoscaler/kubernetes/operator_configs/example_cluster_resource.yaml [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+          # This volume allocates shared memory for Ray to use for its plasma
+          # object store. If you do not provide this, Ray will fall back to
+          # /tmp which cause slowdowns if is not a shared memory volume.
+          volumeMounts:
+          - mountPath: /dev/shm
+            name: dshm
+          resources:
+            requests:
+              cpu: 1000m
+              memory: 512Mi
```
Comment (yiranwang52) [inline python/ray/autoscaler/kubernetes/operator_configs/example_cluster_resource.yaml:53]: I wonder if it's okay for memory request to be different from memory limit.

File: python/ray/autoscaler/kubernetes/operator_configs/example_cluster_resource.yaml (hunk 1 above)
Comment (edoakes) [reply python/ray/autoscaler/kubernetes/operator_configs/example_cluster_resource.yaml:40]: I think this was just to show how you would do it -- redis uses only a single port by default AFAIK

File: python/ray/autoscaler/kubernetes/operator_configs/example_cluster_resource.yaml (hunk 2 above)
Comment (edoakes) [reply python/ray/autoscaler/kubernetes/operator_configs/example_cluster_resource.yaml:53]: Good question. It's definitely a bit weird to have as the example

File: python/ray/autoscaler/kubernetes/operator_configs/example_cluster_resource.yaml (hunk 1 above)
Comment (Gekho457) [reply python/ray/autoscaler/kubernetes/operator_configs/example_cluster_resource.yaml:40]: Keeping things simple -- going to remove the two extra ports.

**Comment by Gekho457** [comment]
Currently debugging the following error, which occurs after starting a cluster, shutting it down, and trying to start a new cluster. This takes place immediately after the monitor initializes a StandardAutoscaler. 

```[2020-11-29 20:54:33,163 C 26 88] service_based_gcs_client.cc:207: Couldn't reconnect to GCS server. The last attempted GCS server address was :0*** StackTrace Information ***    @     0x7fe5330f1535  google::GetStackTraceToString()    @     0x7fe53306619e  ray::GetCallTrace()    @     0x7fe53308b5a4  ray::RayLog::~RayLog()    @     0x7fe532cd937a  ray::gcs::ServiceBasedGcsClient::ReconnectGcsServer()    @     0x7fe532cd93fd  ray::gcs::ServiceBasedGcsClient::GcsServiceFailureDetected()    @     0x7fe532ce20df  _ZNSt17_Function_handlerIFvRKN3ray6StatusERKNS0_3rpc19GetAllNodeInfoReplyEEZNS4_12GcsRpcClient14GetAllNodeInfoERKNS4_21GetAllNodeInfoRequestERKSt8functionIS8_EEUlS3_S7_E_E9_M_invokeERKSt9_Any_dataS3_S7_    @     0x7fe532ce6f65  ray::rpc::ClientCallImpl<>::OnReplyReceived()    @     0x7fe532bfcfd2  _ZN5boost4asio6detail18completion_handlerIZN3ray3rpc17ClientCallManager29PollEventsFromCompletionQueueEiEUlvE_E11do_completeEPvPNS1_19scheduler_operationERKNS_6system10error_codeEm    @     0x7fe53315eba1  boost::asio::detail::scheduler::do_run_one()    @     0x7fe53315f7f9  boost::asio::detail::scheduler::run()    @     0x7fe533161b57  boost::asio::io_context::run()    @     0x7fe532bc64c4  _ZNSt6thread11_State_implINS_8_InvokerISt5tupleIJZN3ray3gcs19GlobalStateAccessorC4ERKSsS7_bEUlvE_EEEEE6_M_runEv    @     0x7fe5334042a0  execute_native_thread_routine    @     0x7fe534143609  start_thread    @     0x7fe53406a293  clone
--```

**Comment by Gekho457** [comment]
Running the monitor in a subprocess fixed the previous issue -- the code is now functional!
To-do list in PR description.

File: python/ray/operator/operator.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+        self.subprocess = None
+
+    def do_in_subprocess(self, f, wait_to_finish=False):
+        # First stop the subprocess if it's alive
+        if self.subprocess and self.subprocess.is_alive():
+            self.subprocess.terminate()
+            self.subprocess.join()
+        # Reinstantiate process with f as target and start.
+        self.subprocess = mp.Process(name=self.name, target=f)
```
Comment (yiranwang52) [inline python/ray/operator/operator.py:49]: is it possible for us to leak subprocesses if operator.py is killed unexpectedly?

Comment (Gekho457) [reply python/ray/operator/operator.py:49]: I think currently if `operator.py` is killed unexpectedly, the operator pod will shut down.
Which reminds me -- I was going to make all of the ray-clusters managed by the operator fate-share with the operator.
So then everything would go down.
That's of course not optimal behavior.  

Let me also experiment to see what happens if `operator.py` is killed unexpectedly but the pod doesn't go down.

Comment (yiranwang52) [reply python/ray/operator/operator.py:49]: I don't think you want to kill all the clusters when operator go down.
It just need to know about them when started.

Comment (Gekho457) [reply python/ray/operator/operator.py:49]: Ah, right. I think that should work with the code as it is.

Comment (Gekho457) [reply python/ray/operator/operator.py:49]: I think Python does an OK  job of cleaning up processes spawned by the multiprocessing module?
To check, I ran a test script that uses multiprocessing to spawn a dummy process that runs forever. After doing a Ctrl-C keyboard interrupt, the pid of the process is no longer present in the output of `ps -ef`.

I'll check that the monitor processes behave in the same way.

Let me know if there's something that can be done to ensure that the processes are cleaned correctly.
(besides rewriting everything to have each cluster's autoscaler `.update` in a for loop, which is probably a good idea to implement sooner or later or sooner to replace this subprocess logic)

Comment (yiranwang52) [reply python/ray/operator/operator.py:49]: Your test sounds good enough.

**Review by yiranwang52 — APPROVED** [review]

File: python/ray/operator/operator.py (hunk 3 above)
Comment (Gekho457) [reply python/ray/operator/operator.py:49]: Actually, you were right -- terminating `operator.py` did leak the monitor process. (In my test script, the child process was receiving the signal interrupt.)

I've now set the monitor subprocess to be a daemon, and that works -- when I run `kill -SIGTERM <operator_pid>` it stops the monitor subprocess too.
(If you do `kill -SIGKILL <operator_pid>` then the daemon monitor subprocess leaks.)

Comment (yiranwang52) [reply python/ray/operator/operator.py:52]: okay then this is still a problem we should fix in the near future.

**Pull request merged**